### ISSUE TITLE
Warn when defaulting through the substitution

### ISSUE THREAD

**Issue by yav** [GaloisInc/cryptol#536]
When we are done with type inference, we look for any left-over type variables, and just pick some type for them (currently `BIt` or `0` depending on the kind).   This shouldn't really affect the workings of the program as nothing forced these variables to be one thing or another, and they were not generalized, perhaps because they are part of something that can only have a monomorphic type (e.g., variables bound by pattern bindings).

Currently we don't generate warnings when we do this defaulting, but it will be more consistent to do so, as in some cases the results could be a bit confusing.   For example:

    (x,y) = error "blarg"

in this case `x` and `y` will both end up being defaulted to `Bit` unless something else in the module forces them to have some other type.

**Comment by robdockins** [comment]
I've just verified that this example still has the behavior described here with the new defaulting system. What should we do with this one? Our new philosophy is basically "only default to unlimited precision types", but for this case of totally unconstrained variables, I'm not sure what we want.

**Comment by yav** [comment]
Yeah, that's a whole separate kind of defaulting, and I am not sure what would be best to do.  Some options:
  * Do nothing, and leave it as it (i.e., not warning)
  * Add a warning
  * Make it an error, and reject such programs, requiring that users specify the types explicitly.

As far as I know such programs are sufficiently rare that I am not sure if it matters what we do.

**Comment by brianhuffman** [comment]
For totally unconstrained type variables, a parametricity argument says that it doesn't matter how they're instantiated, isn't that right? Does Cryptol have the appropriate parametricity property to make this argument work? (I think that before we added the `Logic` and `Zero` type constraints, it definitely didn't; but it might now.)

**Comment by robdockins** [comment]
From a semantics point of view, I think you're right, it shouldn't make any difference.

I think the more important question is the UX question. It seems to me like this situation would only arise from the user writing something unintended, so maybe it is worth emitting a warning?
